frontend/projects/upgrade/src/app/core/stratification-factors/store/stratification-factors.effects.ts

```diff
@@ -0,0 +1,105 @@
+  ) {}
+
+  fetchStratificationFactors$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionFetchStratificationFactors),
+      withLatestFrom(this.store$.pipe(select(selectAllStratificationFactors))),
+      switchMap(() =>
+        this.stratificationFactorsDataService.fetchStratificationFactors().pipe(
+          map((data: any) =>
+            StratificationFactorsActions.actionFetchStratificationFactorsSuccess({
+              stratificationFactors: data,
+            })
+          ),
+          catchError(() => [StratificationFactorsActions.actionFetchStratificationFactorsFailure()])
+        )
+      )
+    )
+  );
+
+  deleteStratificationFactor$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionDeleteStratificationFactor),
+      map((action) => action.factor),
+      filter((factor) => !!factor),
+      switchMap((factor) =>
+        this.stratificationFactorsDataService.deleteStratificationFactor(factor).pipe(
+          map((data: any) => {
+            this.router.navigate(['/participants']);
+            const stratificationFactor = {
+              ...data[0],
+              factor: data[0].stratificationFactorName,
+            };
+            return StratificationFactorsActions.actionDeleteStratificationFactorSuccess({
+              stratificationFactor: stratificationFactor,
+            });
+          }),
+          catchError(() => [StratificationFactorsActions.actionDeleteStratificationFactorFailure()])
+        )
+      )
+    )
+  );
+
+  importStratificationFactor$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionImportStratificationFactor),
+      map((action) => ({ csvData: action.csvData })),
+      filter(({ csvData }) => !!csvData),
+      switchMap(({ csvData }) =>
+        this.stratificationFactorsDataService.importStratificationFactors(csvData).pipe(
+          map((data) => {
+            return StratificationFactorsActions.actionImportStratificationFactorSuccess();
+          }),
+          catchError(() => [StratificationFactorsActions.actionImportStratificationFactorFailure()])
+        )
+      )
+    )
+  );
+
+  exportStratificationFactor$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionExportStratificationFactor),
+      map((action) => ({ factor: action.factor })),
+      filter(({ factor }) => !!factor),
+      switchMap(({ factor }) =>
+        this.stratificationFactorsDataService.exportStratificationFactor(factor).pipe(
+          map((data) => {
+            this.download(data);
+            return StratificationFactorsActions.actionExportStratificationFactorSuccess();
+          }),
+          catchError(() => [StratificationFactorsActions.actionExportStratificationFactorFailure()])
+        )
+      )
+    )
+  );
+
+  private download(csvData) {
```

Comment:
I know I've seen this kind of "download" function plunked inside the effects file elsewhere in our codebase, but I don't think we should repeat it. The ngrx effect pattern would be to dispatch an action that would prompt the actual component to fire this. Plus we're doing DOM manipulation, that's component business, this file should just be for ngrx effects.

Can you create a 'downloadStratificationCSVData' action, then subscribe to it in the actual component(s) where this action should be received?

Reply:
@danoswaltCL Okay, so can we move the download logic and DOM manipulations in the service and call that instead?

Reply:
I'm gonna take a deeper look at this when I get back in tomorrow, I want to give you a good recommendation. The reason this looks extra strange to me is that we're not really using the success and failure import / export actions for anything. it doesn't touch the reducer or the state, we're just firing a DOM event and we don't do anything with the logic in the effect, so as it is currently written, it doesn't even make sense to dispatch an action, the download action could happen in the component by itself.

Reply:
Okay, let's keep it as it is for now as we discussed on slack. We can later create a separate enhancement PR to hadle this for all download related components at once.